File: apps/android/app/src/main/kotlin/dev/hyphen/android/transport/ProtocolSession.kt
package dev.hyphen.android.transport

import java.io.IOException
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong
import javax.net.ssl.SSLSocket
import kotlin.concurrent.thread

/**
 * Steady-state protocol session over an authenticated TLS socket
 * (HYP-M2-012, protocol v0 §4): heartbeats every interval, receive-
 * silence watchdog (degraded after 2 missed intervals), automatic acks
 * for `requiresAck` envelopes, ack-timeout detection, per-sender seq.
 * Hello/capability negotiation and reconnect/resume land with M2-013.
 */
class ProtocolSession(
    private val socket: SSLSocket,
    private val sessionId: String,
    private val config: Config = Config(),
    private val listener: Listener,
) {
    data class Config(
        /** Protocol §4 default 10 s; tests shrink it. */
        val heartbeatIntervalMs: Long = 10_000,
        val missThreshold: Int = 2,
        val ackTimeoutMs: Long = 10_000,
        /** Test hook only: false simulates a protocol-violating silent peer. */
        val autoAck: Boolean = true,
        /** Seq already consumed on this connection (1 after a hello). */
        val startingSeq: Long = 0,
    )

    interface Listener {
        /** Plugin/feature envelopes; core types (heartbeat/ack) stay internal. */
        fun onEnvelope(envelope: Envelope) {}
        fun onLiveness(state: HeartbeatMonitor.State) {}
        /** A tracked requiresAck envelope was acknowledged by the peer. */
        fun onAck(messageId: String) {}
        fun onAckTimeout(messageId: String) {}
        fun onProtocolError(code: String, detail: String) {}
        fun onClosed() {}
    }

    private val seq = AtomicLong(config.startingSeq)
    private val closed = AtomicBoolean(false)
    private val writeLock = Any()
    private var nextInboundSeq = config.startingSeq + 1
    private lateinit var scheduler: ScheduledExecutorService

    private val monitor = HeartbeatMonitor(
        intervalMs = config.heartbeatIntervalMs,
        missThreshold = config.missThreshold,
        startedAtMs = monotonicNowMs(),
        onStateChange = listener::onLiveness,
    )
    private val ackTracker = AckTracker(config.ackTimeoutMs, listener::onAckTimeout)

    @Synchronized
    fun start() {
        if (closed.get()) return
        // Read deadline at 2x the degraded threshold: with live heartbeats
        // it never trips; on a truly dead link the reader wakes itself up.
        // It also bounds a JSSE hazard — SSLSocket.close() from another
        // thread can block until a concurrent blocking read times out, so
        // an unbounded read would make stop() hang and the peer never see
        // the FIN (found via the M2-013 reconnect integration test).
        runCatching {
            socket.soTimeout = (config.heartbeatIntervalMs * (config.missThreshold + 2))
                .coerceAtMost(Int.MAX_VALUE.toLong()).toInt()
        }
        scheduler = Executors.newSingleThreadScheduledExecutor { runnable ->
            Thread(runnable, "hyphen-session-timer").apply { isDaemon = true }
        }
        scheduler.scheduleAtFixedRate(
            { runCatching { send(Envelope.TYPE_HEARTBEAT) } },
            config.heartbeatIntervalMs, config.heartbeatIntervalMs, TimeUnit.MILLISECONDS,
        )
        val tickMs = (config.heartbeatIntervalMs / 2).coerceAtLeast(1)
        scheduler.scheduleAtFixedRate(
            {
                val now = monotonicNowMs()
                monitor.tick(now)
                ackTracker.tick(now)
            },
            tickMs, tickMs, TimeUnit.MILLISECONDS,
        )
        thread(isDaemon = true, name = "hyphen-session-reader") { readLoop() }
    }

    /** @return the sent messageId (registered for ack tracking if required). */
    fun send(
        type: String,
        payload: Json.Obj = Json.Obj.EMPTY,
        requiresAck: Boolean = false,
        capability: String? = null,
        ackOf: String? = null,
    ): String {
        val wallNow = wallNowMs()
        val timeoutNow = monotonicNowMs()
        val envelope = Envelope(
            messageId = Ulid.generate(wallNow),
            sessionId = sessionId,
            type = type,
            capability = capability,
            seq = seq.incrementAndGet(),
            ackOf = ackOf,
            sentAtUnixMs = wallNow,
            requiresAck = requiresAck,
            payload = payload,
        )
        val bytes = envelope.encode()
        if (requiresAck) ackTracker.registerSent(envelope.messageId, timeoutNow)
        try {
            synchronized(writeLock) { FrameIO.write(socket.outputStream, bytes) }
        } catch (e: IOException) {
            if (requiresAck) ackTracker.unregisterSent(envelope.messageId)
            close()
            throw e
        }
        return envelope.messageId
    }

    fun stop() = close()

    private fun readLoop() {
        while (!closed.get()) {
            val frame = try {
                FrameIO.read(socket.inputStream) ?: break
            } catch (e: FrameIO.FrameTooLarge) {
                listener.onProtocolError("transport/frame-too-large", e.message ?: "")
                break
            } catch (_: IOException) {
                break
            }
            val envelope = try {
                Envelope.decode(frame)
            } catch (e: EnvelopeException) {
                listener.onProtocolError("protocol/invalid-envelope", e.message ?: "")
                break
            }
            if (!validateSessionAndSeq(envelope)) break
            monitor.envelopeReceived(monotonicNowMs())
            when (envelope.type) {
                Envelope.TYPE_ACK -> envelope.ackOf?.let { if (ackTracker.ackReceived(it)) listener.onAck(it) }
                Envelope.TYPE_HEARTBEAT -> Unit // liveness already recorded
                else -> {
                    if (envelope.requiresAck && config.autoAck) {
                        runCatching { send(Envelope.TYPE_ACK, ackOf = envelope.messageId) }
                    }
                    listener.onEnvelope(envelope)
                }
            }
        }
        close()
    }

    private fun validateSessionAndSeq(envelope: Envelope): Boolean {
        if (envelope.type == Envelope.TYPE_HELLO) {
            listener.onProtocolError("protocol/invalid-envelope", "hello is only valid during handshake")
            return false
        }
        if (envelope.sessionId != sessionId) {
            listener.onProtocolError("protocol/invalid-envelope", "sessionId does not match active session")
            return false
        }
        if (envelope.seq != nextInboundSeq) {
            listener.onProtocolError(
                "protocol/invalid-envelope",
                "expected seq $nextInboundSeq, got ${envelope.seq}",
            )
            return false
        }
        nextInboundSeq += 1
        return true
    }

    private fun monotonicNowMs(): Long = TimeUnit.NANOSECONDS.toMillis(System.nanoTime())

    private fun wallNowMs(): Long = System.currentTimeMillis()

    @Synchronized
    private fun close() {
        if (closed.getAndSet(true)) return
        if (::scheduler.isInitialized) scheduler.shutdownNow()
        // onClosed reports terminal session state; socket close is best-effort and may block in JSSE.
        listener.onClosed()
        runCatching { socket.close() }
    }
}

File: apps/macos/Sources/HyphenTransport/ProtocolSession.swift
import Foundation
import Network

/// Steady-state protocol session over an authenticated NWConnection
/// (HYP-M2-012, protocol v0 §4): heartbeats every interval, receive-
/// silence watchdog (degraded after 2 missed intervals), automatic acks
/// for `requiresAck` envelopes, ack-timeout detection, per-sender seq.
/// Hello/capability negotiation and reconnect/resume land with M2-013.
/// Twin of the Android `ProtocolSession`.
public final class ProtocolSession {

    public struct Config {
        /// Protocol §4 default 10 s; tests shrink it.
        public var heartbeatIntervalMs: Int64 = 10_000
        public var missThreshold: Int64 = 2
        public var ackTimeoutMs: Int64 = 10_000
        /// Test hook only: false simulates a protocol-violating silent peer.
        public var autoAck = true
        public init() {}
    }

    public struct Callbacks {
        /// Plugin/feature envelopes; core types (heartbeat/ack) stay internal.
        public var onEnvelope: (Envelope) -> Void = { _ in }
        public var onLiveness: (HeartbeatMonitor.State) -> Void = { _ in }
        public var onAckTimeout: (String) -> Void = { _ in }
        public var onProtocolError: (String, String) -> Void = { _, _ in }
        public var onClosed: () -> Void = {}
        public init() {}
    }

    private let connection: NWConnection
    private let sessionId: String
    private let config: Config
    private let callbacks: Callbacks
    private let queue: DispatchQueue
    private let queueKey = DispatchSpecificKey<Bool>()
    private let reader = FrameReader()
    private var monitor: HeartbeatMonitor!
    private var ackTracker: AckTracker!
    private var heartbeatTimer: DispatchSourceTimer?
    private var tickTimer: DispatchSourceTimer?
    private var seq: Int64 = 0
    private var closed = false

    /// - Parameter connection: an already-established (post-handshake)
    ///   connection from `TLSEndpointListener`/`TLSConnector`.
    public init(
        connection: NWConnection,
        sessionId: String,
        config: Config = Config(),
        callbacks: Callbacks
    ) {
        self.connection = connection
        self.sessionId = sessionId
        self.config = config
        self.callbacks = callbacks
        self.queue = DispatchQueue(label: "hyphen-session")
        queue.setSpecific(key: queueKey, value: true)
    }

    public func start() {
        queue.async { [self] in
            monitor = HeartbeatMonitor(
                intervalMs: config.heartbeatIntervalMs,
                missThreshold: config.missThreshold,
                startedAtMs: Self.nowMs(),
                onStateChange: callbacks.onLiveness
            )
            ackTracker = AckTracker(timeoutMs: config.ackTimeoutMs, onTimeout: callbacks.onAckTimeout)

            let heartbeat = DispatchSource.makeTimerSource(queue: queue)
            heartbeat.schedule(
                deadline: .now() + .milliseconds(Int(config.heartbeatIntervalMs)),
                repeating: .milliseconds(Int(config.heartbeatIntervalMs))
            )
            heartbeat.setEventHandler { [weak self] in
                self?.sendOnQueue(type: Envelope.typeHeartbeat)
            }
            heartbeat.resume()
            heartbeatTimer = heartbeat

            let tickMs = max(1, Int(config.heartbeatIntervalMs / 2))
            let tick = DispatchSource.makeTimerSource(queue: queue)
            tick.schedule(deadline: .now() + .milliseconds(tickMs), repeating: .milliseconds(tickMs))
            tick.setEventHandler { [weak self] in
                let now = Self.nowMs()
                self?.monitor.tick(nowMs: now)
                self?.ackTracker.tick(nowMs: now)
            }
            tick.resume()
            tickTimer = tick

            receiveLoop()
        }
    }

    /// Thread-safe; callable from any thread EXCEPT inside this session's
    /// own callbacks (which already run on the session queue).
    @discardableResult
    public func send(
        type: String,
        payload: [String: Any] = [:],
        requiresAck: Bool = false,
        capability: String? = nil,
        ackOf: String? = nil
    ) -> String? {
        if DispatchQueue.getSpecific(key: queueKey) == true {
            return sendOnQueue(type: type, payload: payload, requiresAck: requiresAck, capability: capability, ackOf: ackOf)
        }
        return queue.sync {
            sendOnQueue(type: type, payload: payload, requiresAck: requiresAck, capability: capability, ackOf: ackOf)
        }
    }

    public func stop() {
        queue.async { self.close() }
    }

    @discardableResult
    private func sendOnQueue(
        type: String,
        payload: [String: Any] = [:],
        requiresAck: Bool = false,
        capability: String? = nil,
        ackOf: String? = nil
    ) -> String? {
        guard !closed else { return nil }
        let now = Self.nowMs()
        seq += 1
        let envelope = Envelope(
            messageId: Ulid.generate(),
            sessionId: sessionId,
            type: type,
            capability: capability,
            seq: seq,
            ackOf: ackOf,
            sentAtUnixMs: now,
            requiresAck: requiresAck,
            payload: payload
        )
        do {
            let frame = try FrameCodec.encode(try envelope.encode())
            connection.send(content: frame, completion: .contentProcessed { [weak self] error in
                if error != nil { self?.close() }
            })
        } catch {
            callbacks.onProtocolError("protocol/invalid-envelope", "encode failed: \(error)")
            return nil
        }
        if requiresAck { ackTracker.registerSent(messageId: envelope.messageId, nowMs: now) }
        return envelope.messageId
    }

    private func receiveLoop() {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 65_536) { [weak self] data, _, isComplete, error in
            guard let self, !self.closed else { return }
            if let data, !data.isEmpty {
                do {
                    for frame in try self.reader.feed(data) {
                        self.handleFrame(frame)
                    }
                } catch {
                    self.callbacks.onProtocolError("transport/frame-too-large", "\(error)")
                    self.close()
                    return
                }
            }
            if isComplete || error != nil {
                self.close()
                return
            }
            self.receiveLoop()
        }
    }

    private func handleFrame(_ frame: Data) {
        let envelope: Envelope
        do {
            envelope = try Envelope.decode(frame)
        } catch {
            // Stay open per §3; malformed envelopes are surfaced and skipped.
            callbacks.onProtocolError("protocol/invalid-envelope", "\(error)")
            return
        }
        monitor.envelopeReceived(nowMs: Self.nowMs())
        switch envelope.type {
        case Envelope.typeAck:
            if let ackOf = envelope.ackOf { ackTracker.ackReceived(ackOf) }
        case Envelope.typeHeartbeat:
            break // liveness already recorded
        default:
            if envelope.requiresAck && config.autoAck {
                sendOnQueue(type: Envelope.typeAck, ackOf: envelope.messageId)
            }
            callbacks.onEnvelope(envelope)
        }
    }

    private func close() {
        guard !closed else { return }
        closed = true
        heartbeatTimer?.cancel()
        tickTimer?.cancel()
        connection.cancel()
        callbacks.onClosed()
    }

    private static func nowMs() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }
}
